src/openzaak/components/zaken/api/validators.py

```diff
@@ -280,6 +281,94 @@ def __call__(self, attrs: dict):
                 raise serializers.ValidationError(self.message, code=self.code)
 
 
+class DeelzaakReopenValidator:
+    """
+    Validate that the hoofdzaak is not closed  when a deelzaak is reopened.
+    """
+
+    code = "hoofdzaak-closed"
+    message = _(
+        "Een deelzaak kan alleen worden heropend als de hoofzaak niet afgesloten is."
+    )
+
+    def __call__(self, attrs: dict):
+        if attrs.get("__is_eindstatus"):
+            return
+
+        zaak = attrs.get("zaak")
+        if not zaak or not zaak.hoofdzaak or not zaak.hoofdzaak.current_status:
+            return
+
+        if zaak.hoofdzaak.current_status.statustype.is_eindstatus():
+            raise serializers.ValidationError(self.message, code=self.code)
+
+
+class EndStatusDeelZakenValidator:
+    """
+    Validate that related deelzaken are closed when trying to close a hoofdzaak
+    (by setting the eindstatus on that zaak)
+
+    The serializer sets the __is_eindstatus attribute in the data dict as
+    part of the ``to_internal_value`` method.
+    """
+
+    code = "deelzaken-not-closed"
+    message = _(
+        "Er zijn gerelateerde deelzaken die nog niet zijn afgesloten zijn."
+        "Deze deelzaken moeten eerst afgesloten worden voordat de zaak afgesloten kan worden."
+    )
+
+    def __call__(self, attrs: dict):
+        if not attrs.get("__is_eindstatus"):
+            return
+
+        zaak = attrs.get("zaak")
+        # earlier validation failed possibly
+        if not zaak or not zaak.deelzaken.exists():
+            return
+
+        self.check_status_exists(zaak.deelzaken)
+        self.check_internal_status_types(zaak.deelzaken.filter(_zaaktype__isnull=False))
+        self.check_external_status_types(
+            zaak.deelzaken.filter(_zaaktype_relative_url__isnull=False)
+        )
+
+    def check_status_exists(self, qs):
+        if qs.filter(status__isnull=True).exists():
+            raise serializers.ValidationError(self.message, code=self.code)
+
+    def check_internal_status_types(self, qs):
```

Comment:
I think I'm fine with keeping this query the way it is now, the code would be simplified by #2013, but the underlying SQL isn't actually that complex.

Reply:
Closed #2013